Add RGB LED Strip support

boweeble opened the pull request:
- Use existing M150 in the GCODE
Example: M151 R255 U0 B0 will turn the LED Strip RED

- When printing from the SD Card the LED Strip will turn from BLUE to
VIOLET until the bed is at temperature, then from VIOLET to RED as the
hot end reaches temperature at which point the LED strip will turn white
to illuminate the print area.
- Once the print has completed, the LED Strip will turn GREEN until the
user presses the encoder button.

- Requires LCD & SDCARD support and requires the Servo (3) PWM pins.

boweeble committed:
Add RGB LED Strip support

- Added M151 in the GCODE
Ex: M151 R255 U0 B0 will turn the LED Strip RED

- When printing from the SD Card the LED Strip will turn from BLUE to
YELLOW until the bed is at temperature, then from YELLOW to RED as the
hot end reaches temperature at which point the LED strip will turn white
to illuminate the print area.
- Once the print has completed, the LED Strip will turn green until the
user presses the encoder or a button.

- Requires LCD & SDCARD support and uses the Servo PWM pins

thinkyhead commented on Marlin/Configuration.h:
```diff
@@ -49,7 +49,7 @@
  * Note: Update also Version.h !
  */
-#define CONFIGURATION_H_VERSION 010100
+#define CONFIGURATION_H_VERSION 010101
```
The guidelines for changing the version need to be updated. We'll only be changing it for tagged releases.

boweeble replied:
Could you point me in the right direction, I could not find information on this...

thinkyhead commented on Marlin/Configuration.h:
```diff
@@ -1445,4 +1445,42 @@
+    #if ENABLED(LED_STRIP)
+      #define REDPIN 4
+      #define GREENPIN 5
+      #define BLUEPIN 6
```
These names will need to be more unique. `LED_STRIP_RED_PIN`, etc.

boweeble replied:
As suggested in some other comments, I will be changing this feature to utilize the BLINKM/RGB_LED feature.

thinkyhead commented on Marlin/LED_Strip.cpp:
```diff
@@ -0,0 +1,32 @@
+  LCD_MESSAGEPGM(WELCOME_MSG);
+  lcd_update();
+}
+
```
Please indent by 2 spaces any code between `#if` and `#endif`.

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -132,6 +132,7 @@
  * M145 - Set heatup values for materials on the LCD. H<hotend> B<bed> F<fan speed> for S<material> (0=PLA, 1=ABS)
  * M149 - Set temperature units. (Requires TEMPERATURE_UNITS_SUPPORT)
  * M150 - Set Status LED Color as R<red> U<green> B<blue>. Values 0-255. (Requires BLINKM or RGB_LED)
+ * M151 - Set RGB LED Strip Color Output R<red> U:<green> B<blue>. Values 0-255 (Requires LED_STRIP)
```
Colon required for `U`?

boweeble replied:
Will be removed and utilize the M150 command.

thinkyhead commented on Marlin/LED_Strip.cpp:
```diff
@@ -0,0 +1,32 @@
+void rgb_to_led_strip(int red, int green, int blue) {    
+  analogWrite(REDPIN, red);
+  analogWrite(GREENPIN, green);
+  analogWrite(BLUEPIN, blue);
```
As it's possible to connect to non-PWM pins, and as there is now a general concept of color LEDs for Marlin, please implement this as part of the existing `set_led_color` function instead of creating a new function.

thinkyhead replied:
If this needs to require PWM-able pins, we actually have a way to check for that, which works for any board.

boweeble replied:
This will require PWM-able pins, where do I find documentation to check for this?

thinkyhead replied:
It's not completely settled yet, but we use the technique to check in `pinsDebug.h`. For now we can actually the old "trick" that's been used up to now:
```cpp
digitalWrite(RGB_LED_RED_PIN, r ? HIGH : LOW);
analogWrite(RGB_LED_RED_PIN, r);
```
If the pin isn't analog (hence PWM) then the second command just doesn't do anything. It's already implemented this way in the existing `set_led_color`.

thinkyhead commented on Marlin/LED_Strip.cpp:
```diff
@@ -0,0 +1,32 @@
+}
+
+// Print finished - wait for user click
+void rgb_print_finished() {
```
I consider this to be a separate feature, or at least a sub-feature dependent on having one of the color LED options. Let's look at generalizing this so that it can work with `BLINKM` and `RGB_LED`, and perhaps as an option for `TEMP_STAT_LEDS` also.

boweeble replied:
Will do...

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -5391,6 +5396,20 @@ inline void gcode_M109() {
+    #if ENABLED(LED_STRIP)
+      // Gradually change LED strip from yellow to red as the hotend heats up
+      // Change is calculated from room temperature (20C or 68F) to show a smooth change
+      if(wait_for_heatup) {
```
One space is required between `if` and the opening parenthesis.

boweeble replied:
Fixed

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -5419,7 +5438,14 @@ inline void gcode_M109() {
+    #if ENABLED(LED_STRIP)
+      // Heating is complete, turn the LED Strip to all white
+      rgb_to_led_strip(255, 255, 255);
+    #endif
```
If generalized for all the LED options, then this will be:
```cpp
    #if ENABLED(PRINTER_EVENT_LEDS)
      // Heating is complete, turn the status light to all white
      set_led_color(255, 255, 255);
    #endif
```

boweeble replied:
Great idea and makes more sense.

boweeble committed:
Update Marlin_main.cpp

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -5510,6 +5536,19 @@ inline void gcode_M109() {
+          index_multiplier = 255 / (theTarget - 20);
+          r = 255 * (temp / index_multiplier);
+          g = r;
+          b = 255 - (temp * index_multiplier);
```
Isn't blue off for yellow? And green should then decrease, while red should be always-on…?
```cpp
          r = 255;
          g = 255 - map(temp, 20, theTarget, 0, 255);
          b = 0;
```

thinkyhead replied:
Ah, never mind, I see you want to go from BLUE to YELLOW for the bed. Interesting! I still recommend using the `map()` function. It's designed for just this kind of situation.

boweeble replied:
Changing this to go from Blue to Violet then Violet to Red...makes sense to implement the map() function as it's more efficient and cleaner.

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -6045,6 +6084,29 @@ inline void gcode_M121() { endstops.enable_globally(false); }
+    code_seen('R') ? (code_has_value() ? code_value_byte() : 255) : 0,
+    code_seen('U') ? (code_has_value() ? code_value_byte() : 255) : 0,
+    code_seen('B') ? (code_has_value() ? code_value_byte() : 255) : 0
+  );
```
Oops. This is missing the function call.

You can actually overload `M150` instead of adding `M151`.

boweeble replied:
Will utilize M150

thinkyhead commented:
`Marlin_main.cpp` got a little mangled. But you probably saw that already!

Tannoo commented:
Would it be possible to also use this feature when not printing from SDcard?

Tannoo commented:
And again about an active low option.

Tannoo commented:
And again about an active low option.

thinkyhead commented:
This needs a rebase on the latest `RCBugFix`, and conflict-resolution.

boelle commented:
@boweeble great idea :-D as @thinkyhead says it need a rebase and conflict fix

what led strips do you use?

thinkyhead commented:
In the future we'll make it easier to have multiple lights, such as one inside your LCD controller (currently `BLINKM`), plus an LED strip (`LED_STRIP`) or single RGB LED (`RGB_LED`). We'll just come up with a way to refer to the particular one you want to set (most likely a `P` parameter in `M150`).

ghost commented:
@boelle,
I don't know which one @boweeble used.

I got this https://www.sparkfun.com/products/12022 and some TIP220s to use for it.

Now, just waiting on the merge. Or at least a PR that has the corrections mentioned.

boweeble commented:
@boelle, @Tannoo I used this one:

1 - 12V Flexible SMD 5050 RGB LED Strip Lights
https://www.amazon.com/gp/product/B006LW2NJM

3 - MOSFET N-Ch 60 Volt 16 Amp (which is a bit of an overkill)
https://www.amazon.com/gp/product/B011NB16WK

Will be reworking this and hopefully have something ready to test first part of next week.  Thanks for the interest!

I will provide a wiring diagram as well...

ghost commented:
Overkill? There is no such thing. The TIP220s are rated for 5A and are NPN.

boelle commented:
@boweeble there are a file conflict in Marlin_main.cpp

boweeble committed:
Add RGB LED Strip support

Added RGB_STRIP define
Added PRINTER_EVENT_LEDS

- RGB_STRIP
- Use existing M150 in the GCODE
Example: M151 R255 U0 B0 will turn the LED Strip RED

- PRINTER_EVENT_LEDS
- LEDs will turn from BLUE to
VIOLET until the bed is at temperature
- LEDs will turn from VIOLET to RED as the
hot end reaches temperature
- When printer is at temp and ready to print LEDs will turn white
to illuminate the print area.
- LEDs will turn GREEN when print is finished until the
user presses the encoder button.

- Requires the Servo (3) PWM pins.

boweeble commented:
The change is actually fairly small/minor but I re-based with the RC8 that just came out.

ghost commented:
I got this implemented on mine and it works! Thank you @boweeble!

ghost commented:
#####     I'm still trying to get the hang of GitHub desktop. So, this will have to suffice for now. This is a rework of the last print_event. It will timeout waiting for the user.

`  void handle_led_print_event(int code) {`

    byte wait_to_reset = 600;   // 600 = 60 secs.
        
    switch(code) {
      case(0):        // Print Complete
        LCD_MESSAGEPGM(MSG_INFO_COMPLETED_PRINTS);
        lcd_update();
        set_rgb_color(0, 255, 0);  // Turn RGB LEDs to GREEN
        wait_for_user = true;
        byte wait_for_user_timeout;

        #if ENABLED(ULTRA_LCD)
          wait_for_user_timeout = 0;
          do {
           lcd_update();
           safe_delay(100);
           wait_for_user_timeout ++;
           if (wait_for_user_timeout >= wait_to_reset) break;
          } while (wait_for_user);
        #endif // ULTRA_LCD

        #if ENABLED(DOGLCD)
          wait_for_user_timeout = 0;
          do {
           lcd_update();
           safe_delay(100);
           wait_for_user_timeout ++;
           if (wait_for_user_timeout >= wait_to_reset) break; //
          } while (wait_for_user);
        #endif // DOGLCD

        wait_for_user = false;
        set_rgb_color(0, 0, 0);  // Turn RGB LEDs off

        LCD_MESSAGEPGM(WELCOME_MSG);
        lcd_update();
        break;

      case(1):      // Printer Ready for Printing (All WHITE)
        set_rgb_color(255, 255, 255);
        break;
        
    } // switch(code)

`}
`

ghost commented:
I also put this at the end of M84 function in Marlin_main.cpp (~5956):

`  #if ENABLED(BLINKM) || ENABLED(RGB_LED) || ENABLED(RGB_STRIP) && ENABLED(PRINTER_EVENT_LEDS)
    if (code_value_temp_abs() <= (EXTRUDE_MINTEMP)/2) handle_led_print_event(0);
  #endif
`

TylonHH commented:
Errr, where do I connect the LED Stripe?
